include/xtensor/xsemantic.hpp

```diff
@@ -18,8 +18,211 @@
+            static bool check_tuple(const std::tuple<T...>& t, const memory_range& dst_range)
+            {
+                using ChildE = std::decay_t<decltype(std::get<I>(t))>;
+                return overlapping_memory_checker_traits<ChildE>::check_overlap(std::get<I>(t), dst_range) || check_tuple<I + 1>(t, dst_range);
+            }
+
+            static bool check_overlap(const E& expr, const memory_range& dst_range)
+            {
+                if (expr.size() == 0)
+                {
+                    return false;
+                }
+                else
+                {
+                    return check_tuple(expr.arguments(), dst_range);
+                }
+            }
+        };
+
+        template <class E>
+        struct overlapping_memory_checker_traits<E, std::enable_if_t<!has_memory_address<E>::value &&
+                                                                    is_specialization_of<xgenerator, E>::value>>
+        {
+            static bool check_overlap(const E&, const memory_range&)
+            {
+                return false;
+            }
+        };
+
+        struct overlapping_memory_checker_base
+        {
+            memory_range m_dst_range;
+
+            explicit overlapping_memory_checker_base() = default;
+            explicit overlapping_memory_checker_base(memory_range dst_memory_range) : m_dst_range(std::move(dst_memory_range)) {}
+
+            template <class E>
+            bool operator()(const E& expr) const
+            {
+                if (!m_dst_range.m_first || !m_dst_range.m_last)
+                {
+                    return false;
+                }
+                else
+                {
+                    return overlapping_memory_checker_traits<E>::check_overlap(expr, m_dst_range);
+                }
+            }
+        };
+
+        template <class Dst, class Enable = void>
+        struct overlapping_memory_checker : overlapping_memory_checker_base
+        {
+            explicit overlapping_memory_checker(const Dst&) : overlapping_memory_checker_base() {}
+        };
+
+        template <class Dst>
+        struct overlapping_memory_checker<Dst, std::enable_if_t<has_memory_address<Dst>::value>> : overlapping_memory_checker_base
+        {
+            explicit overlapping_memory_checker(const Dst& aDst)
+                : overlapping_memory_checker_base([&]() {
+                      if (aDst.size() == 0)
+                      {
+                          return memory_range();
+                      }
+                      else
+                      {
+                          return memory_range(std::addressof(*aDst.begin()), std::addressof(*aDst.rbegin()));
+                      }
+                  }())
+            {
+            }
+        };
+
+        template <class Dst>
+        auto make_overlapping_memory_checker(const Dst& a_dst)
+        {
+            return overlapping_memory_checker<Dst>(a_dst);
+        }
```

Comment:
The `overlapping_memory_checker` definitions should go into the file containing the `overlapping_memory_checker_traits` structs when they're moved.